st2common/st2common/util/output_schema.py

```diff
@@ -77,6 +79,61 @@ def mask_secret_output(ac_ex, output_value):
 
 def validate_output(runner_schema, action_schema, result, status, output_key):
     """Validate output of action with runner and action schema."""
+
+    if (
+        output_key == "result"
+        and result["result"]
+        and isinstance(result["result"], dict)
+    ):
+        params_list = result["result"].keys()
+        for param in params_list:
+            if (
+                isinstance(result["result"][param], str)
+                and action_schema[param]["type"] == "string"
+            ):
+                continue
+            elif (
+                isinstance(result["result"][param], dict)
+                and action_schema[param]["type"] == "object"
+            ):
+                continue
+            elif (
+                isinstance(result["result"][param], list)
+                and action_schema[param]["type"] == "array"
+            ):
+                continue
+            elif (
+                isinstance(result["result"][param], int)
+                and action_schema[param]["type"] == "integer"
+            ):
+                continue
+            elif (
+                isinstance(result["result"][param], numbers.Number)
+                and action_schema[param]["type"] == "number"
+            ):
+                continue
+            elif (
+                isinstance(result["result"][param], bool)
+                and action_schema[param]["type"] == "boolean"
+            ):
+                continue
+            elif (
+                result["result"][param] is None
+                and action_schema[param]["type"] == "null"
+            ):
+                continue
+            else:
+                status = action_constants.LIVEACTION_STATUS_FAILED
+                msg = "Failed to validate action output. '{0}' is not of type '{1}' in entry point file.".format(
+                    param, action_schema[param]["type"]
+                )
+                LOG.exception("Failed to validate output. %s", msg)
+                result["exit_code"] = FAILURE_EXIT_CODE
+                result["result"] = "None"
+                result["stderr"] = msg
+                result["stdout"] = ""
+                return (result, status)
+
```

Comment:
This doesn't look right. How come the schema validation isn't able to pick up the type error and we have to manually check this with a bunch of if/elif statements? Can you research and provide a reasonable explanation here?

Reply:
<img width="815" alt="2021-08-05 (8)" src="https://user-images.githubusercontent.com/81608498/128306741-c6f1d640-8f68-4e01-99a7-88e6dc05f487.png">
<img width="832" alt="2021-08-05 (9)" src="https://user-images.githubusercontent.com/81608498/128306769-4775fad5-e9de-477a-9f3d-e1d908c28c48.png">
<img width="960" alt="2021-08-05 (4)" src="https://user-images.githubusercontent.com/81608498/128306785-f3b0f4d5-7428-4938-804a-165aeb9c40c7.png">

Reply:
<img width="960" alt="2021-08-05 (12)" src="https://user-images.githubusercontent.com/81608498/128311602-b358b880-6648-4d33-afcf-2da6c8f9dc4f.png">
<img width="960" alt="2021-08-05 (11)" src="https://user-images.githubusercontent.com/81608498/128311610-b35b2f5d-fd73-4d4b-9ae9-9e2f314d0d4e.png">
<img width="960" alt="2021-08-05 (10)" src="https://user-images.githubusercontent.com/81608498/128311618-7f05b225-1689-4cc9-970c-a686dbcdef08.png">
<img width="960" alt="2021-08-05 (13)" src="https://user-images.githubusercontent.com/81608498/128311624-8b3cd6aa-0758-4ac4-88fe-9997abc3a6e5.png">

Reply:
@m4dcoder As you said schema validation gives type error and gives message why output schema validation failed at CLI but when we run same action from web UI, it only gives `// Action produced no data` as output. As there is no error message, it isn't clear why action failed. After adding conditions for checking output schema property types in metadata and entry point files we can prepare appropriate result if any condition fails. It is giving proper error message at CLI and web UI as well.

Reply:
I have observed that these changes are causing key error in st2 pack install workflow. Please give me inputs for better approach to resolve this issue.

Reply:
The if/else here is just wrong and heading in the wrong direction. The schema validation is doing the job of detecting the type error.  I would check the result with the error and message. Each runners have different format/structure for their result. I think the UI is expecting the error to be in a specific path in the result. Can you check this with the UI developer?  To fix the issue, you may have to make changes at https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/output_schema.py#L95 to write the result correctly for the runner of the action.

Reply:
The if/else here is just wrong and heading in the wrong direction. The schema validation is doing the job of detecting the type error. I would check the result with the error and message. Each runners have different format/structure for their result. I think the UI is expecting the error to be in a specific path in the result. Can you check this with the UI developer? To fix the issue, you may have to make changes at https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/output_schema.py#L95 to write the result correctly for the runner of the action.